agent/xdsv2/rbac_resources.go

```diff
@@ -169,6 +168,197 @@ func makeRBACPolicy(p *pbproxystate.Permission) *envoy_rbac_v3.Policy {
 	}
 }
 
+func l4PolicyLabel(perms []*pbproxystate.Permission, i int) string {
+	if len(perms) == 1 {
+		return baseL4PermissionKey
+	}
+	return fmt.Sprintf("%s-%d", baseL4PermissionKey, i)
+}
+
+func makeL7RBACPolicies(l7Permissions []*pbproxystate.Permission) map[string]*envoy_rbac_v3.Policy {
+	// sort permissions into those with L7-specific features and those without, to match labeling and behavior
+	// conventions in V1: https://github.com/hashicorp/consul/blob/4e451f23584473a7eaf7f123145ca85e0a31783a/agent/xds/rbac.go#L647
+	// this is a somewhat unfortunate carry-over needed for testing v1 vs v2 final config
+	// and this will break with v1 intentions when multiple L4 permissions are used
+	var l4Perms []*pbproxystate.Permission
+	var l7Perms []*pbproxystate.Permission
+	for _, p := range l7Permissions {
+		if len(p.DestinationRules) > 0 {
+			l7Perms = append(l7Perms, p)
+		} else {
+			l4Perms = append(l4Perms, p)
+		}
+	}
+
+	policies := make(map[string]*envoy_rbac_v3.Policy, len(l7Permissions))
+
+	// L7 policies first, then L4 per: https://github.com/hashicorp/consul/blob/4e451f23584473a7eaf7f123145ca85e0a31783a/agent/xds/rbac.go#L664
+	for i, permission := range l7Perms {
+		policy := makeL7RBACPolicy(permission)
+		if policy != nil {
+			policies[fmt.Sprintf("%s-%d", baseL7PermissionKey, i)] = policy
+		}
+	}
+	for i, permission := range l4Perms {
+		policy := makeL4RBACPolicy(permission)
+		if policy != nil {
+			policies[l4PolicyLabel(l4Perms, i)] = policy
+		}
+	}
+
+	return policies
+}
+
+func makeL7RBACPolicy(p *pbproxystate.Permission) *envoy_rbac_v3.Policy {
+	if p == nil || len(p.Principals) == 0 {
+		return nil
+	}
+
+	var principals []*envoy_rbac_v3.Principal
+
+	for _, p := range p.Principals {
+		principals = append(principals, toEnvoyPrincipal(p))
+	}
+	permissions := permissionsFromDestinationRules(p.DestinationRules)
+	return &envoy_rbac_v3.Policy{
+		Principals:  principals,
+		Permissions: permissions,
+	}
+}
+
+func permissionsFromDestinationRules(drs []*pbproxystate.DestinationRule) []*envoy_rbac_v3.Permission {
+	var perms []*envoy_rbac_v3.Permission
+
+	for _, dr := range drs {
```

Comment:
How does this handle the case where a destination rule specifies a path, methods, and headers? It looks like this uses `or`, but I was expecting `and`.